File: go/300.500/384.ShuffleAnArray.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

type Solution struct {
	Orinum  []int
	Randnum []int
}

func init() {
	rand.Seed(time.Now().UnixNano())
}

func Constructor(nums []int) Solution {
	ret := Solution{
		Orinum:  nums,
		Randnum: make([]int, len(nums)),
	}
	copy(ret.Randnum, nums)
	return ret
}

func (this *Solution) Reset() []int {
	return this.Orinum
}

func (this *Solution) Shuffle() []int {
	n := len(this.Randnum)
	for i := 0; i < n; i++ {
		p := rand.Intn(n-i) + i
		this.Randnum[i], this.Randnum[p] = this.Randnum[p], this.Randnum[i]
	}
	return this.Randnum
}

func main() {
	cls := Constructor([]int{1, 2, 3})
	fmt.Println(cls.Shuffle())
	fmt.Println(cls.Reset())
	fmt.Println(cls.Shuffle())
}

File: 300.500/384.ShuffleAnArray.cpp
#include <iostream>
#include <vector>
#include <time.h>
#include <stdlib.h>
using namespace std;

class ShuffleAnArray{
public:
    ShuffleAnArray(vector<int>& nums) {
        orinum = nums;
        randnum = nums;
        srand(time(nullptr));
    }
    
    vector<int> reset() {
        return orinum;
    }
    
    vector<int> shuffle() {
        for(int i = 0; i< randnum.size(); ++i){
            int p = rand()%(randnum.size()-i)+i;
            swap(randnum[i], randnum[p]);
        }
        return randnum;
    }
private:
    vector<int> orinum;
    vector<int> randnum;
};

int main(int argc, char const *argv[]){
    ShuffleAnArray cls(vector<int>{1,2,3});
    cls.shuffle();
    cls.reset();
    cls.shuffle();
    return 0;
}
